### PR TITLE
Fix DiscreteFactorGraph::optimize to return MPE

### LINKED ISSUE #1048: optimize <> MPE
## Description

The `optimize` methods in `DiscreteFactorGraph` and `DiscreteBayesNet` do not actually compute the MPE, but optimize first the marginal of the parents, and then the conditional marginals of the children. While often agreeing with the MPE (I had to work a bit to find a counterexample in addition to the Darwiche example) this computation is not the same, and in fact is not a common thing to do (as opposed to max-marginal, which optimizes all individual marginals). 

This will probably of interest to @keevindoherty , @varunagrawal , @ProfFan 

## Steps to reproduce

I created a counter-example in a branch where I am fixing the issue.

```c++
TEST(DiscreteFactorGraph, marginalIsNotMPE) {
  // Declare 2 keys
  DiscreteKey A(0, 2), B(1, 2);

  // Create Bayes net such that marginal on A is bigger for 0 than 1, but the
  // MPE does not have A=0.
  DiscreteBayesNet bayesNet;
  bayesNet.add(B | A = "1/1 1/2");
  bayesNet.add(A % "10/9");

  // The expected MPE is A=1, B=1
  DiscreteValues mpe;
  insert(mpe)(0, 1)(1, 1);

  // Which we verify using max-product:
  DiscreteFactorGraph graph(bayesNet);
  auto actualMPE = graph.optimize();
  EXPECT(assert_equal(mpe, actualMPE));
  EXPECT_DOUBLES_EQUAL(0.315789, graph(mpe), 1e-5);  // regression

  // Optimize on BayesNet maximizes marginal, then the conditional marginals:
  auto notOptimal = bayesNet.optimize();
  EXPECT(graph(notOptimal) < graph(mpe));
  EXPECT_DOUBLES_EQUAL(0.263158, graph(notOptimal), 1e-5);  // regression
}
```

## Expected behavior

`DiscreteFactorGraph::optimize` should compute the MPE.

## Additional information

I already fixed what optimize does. To maintain the API and the common use case, I will leave the eliminate methods as they are, producing a BayesNet, but will deprecated BayesNet optimize. There will be new maxProduct methods in DFG.

### PR DESCRIPTION
**Pull request by dellaert**
## Major Changes

This PR fixes #1048. In particular:
- `DiscreteFactorGraph::optimize` now really computes the MPE (maximum probable explanation)
- added a method `DiscreteFactorGraph::maxProduct`
- created two versions of each to take `OrderingType` or `Ordering`
- deprecated `DiscreteBayesNet::optimize` and `DiscreteConditional::solve`, as they do nothing really useful
- removed optimalAssignment in `gtsam_unstable/discrete`: redundant with optimize. Not deprecated, just gone!

## Max-product details:
I introduce two new classes (in one header/cpp):
- `DiscreteLookupTable`: inherits from `DiscreteConditional` to play well with generic elimination, but is supposed to be a lookup table of the maximum values obtained during max-product
- `DiscreteLookupDAG`: a DAG of lookup tables. Made this separate from BayesNet as to not confuse.
- `maxProduct` is implemented with a custom `Eliminate` function in `DiscreteFactorGraph.cpp`

### CODE REVIEW

**Comment by dellaert** [comment]
@keevindoherty FYI. If you relied optimize, there might be discrepancies with the actual MPE.

File: gtsam/discrete/DiscreteBayesNet.h [hunk 1]
```diff
@@ -31,12 +31,12 @@
...
 namespace gtsam {
 
-/** A Bayes net made from linear-Discrete densities */
+/** A Bayes net made from discrete conditional distributions. */
   class GTSAM_EXPORT DiscreteBayesNet: public BayesNet<DiscreteConditional>
   {
   public:
 
-    typedef FactorGraph<DiscreteConditional> Base;
+    typedef BayesNet<DiscreteConditional> Base;
```
Comment (varunagrawal) [inline gtsam/discrete/DiscreteBayesNet.h:39]: Oh wow this has been hidden here for how long?

File: gtsam/discrete/DiscreteConditional.cpp [hunk 2]
```diff
@@ -248,59 +247,59 @@ void DiscreteConditional::solveInPlace(DiscreteValues* values) const {
   // Get all Possible Configurations
   const auto allPosbValues = frontalAssignments();
 
-  // Find the MPE
+  // Find the maximum
```
Comment (varunagrawal) [inline gtsam/discrete/DiscreteConditional.cpp:250]: Should this read `Find the maximum probability given assignment`?

File: gtsam/discrete/DiscreteConditional.cpp [hunk 3]
```diff
@@ -323,7 +322,8 @@ size_t DiscreteConditional::sample(const DiscreteValues& parentsValues) const {
   return distribution(rng);
 }
 
-/* ******************************************************************************** */
+/* ********************************************************************************
+ */
```
Comment (varunagrawal) [inline gtsam/discrete/DiscreteConditional.cpp:326]: Kill some stars? Should be the same as above.

File: gtsam/discrete/DiscreteLookupDAG.h [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+ * Atlanta, Georgia 30332-0415
+ * All Rights Reserved
+ * Authors: Frank Dellaert, et al. (see THANKS for the full author list)
+
+ * See LICENSE for the license information
+
+ * -------------------------------------------------------------------------- */
+
+/**
+ * @file DiscreteLookupDAG.h
```
Comment (varunagrawal) [inline gtsam/discrete/DiscreteLookupDAG.h:13]: Rename this file to `DiscreteLookupTable`?

File: gtsam/discrete/DiscreteLookupDAG.h [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+
+namespace gtsam {
+
+class DiscreteBayesNet;
+
+/**
+ * @brief DiscreteLookupTable table for max-product
+ *
+ * Inherits from discrete conditional for convenience, but is not normalized.
+ * Is used in pax-product algorithm.
```
Comment (varunagrawal) [inline gtsam/discrete/DiscreteLookupDAG.h:37]: `max-product` not `pax-product`

File: gtsam/discrete/DiscreteLookupDAG.cpp [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+ * Atlanta, Georgia 30332-0415
+ * All Rights Reserved
+ * Authors: Frank Dellaert, et al. (see THANKS for the full author list)
+
+ * See LICENSE for the license information
+
+ * -------------------------------------------------------------------------- */
+
+/**
+ *  @file DiscreteLookupTable.cpp
```
Comment (varunagrawal) [inline gtsam/discrete/DiscreteLookupDAG.cpp:13]: Incorrect file name.

File: gtsam/discrete/DiscreteLookupDAG.h [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+
+  /**
+   * @brief argmax by back-substitution, given certain variables.
+   *
+   * Assumes the DAG is reverse topologically sorted *and* that the
+   * DAG does not contain any conditionals for the given variables.
+   *
+   * @return given assignment extended w. optimal assignment for all variables.
+   */
+  DiscreteValues argmax(DiscreteValues given) const;
```
Comment (varunagrawal) [inline gtsam/discrete/DiscreteLookupDAG.h:132]: Provide default and kill the other definition?

File: gtsam/discrete/DiscreteLookupDAG.h [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+  using This = DiscreteLookupDAG;
+  using shared_ptr = boost::shared_ptr<This>;
+
+  /// @name Standard Constructors
+  /// @{
+
+  /// Construct empty DAG.
+  DiscreteLookupDAG() {}
+
+  // Create from BayesNet with LookupTables
```
Comment (varunagrawal) [inline gtsam/discrete/DiscreteLookupDAG.h:88]: `///` not `//`

File: gtsam/discrete/DiscreteFactorGraph.h [hunk 9]
```diff
@@ -128,18 +129,47 @@ class GTSAM_EXPORT DiscreteFactorGraph
...
-//  GTSAM_EXPORT void reduceWithInverse(const internal::Reduction& inverseReduction);
+  /**
+   * @brief Find the maximum probable explanation (MPE) by doing max-product.
+   *
+   * @param ordering
+   * @return DiscreteValues : MPE
+   */
+  DiscreteValues optimize(const Ordering& ordering) const;
+
+  //  /** Permute the variables in the factors */
```
Comment (varunagrawal) [inline gtsam/discrete/DiscreteFactorGraph.h:166]: Kill?

File: gtsam/discrete/tests/testDiscreteLookupDAG.cpp [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+ * Atlanta, Georgia 30332-0415
+ * All Rights Reserved
+ * Authors: Frank Dellaert, et al. (see THANKS for the full author list)
+
+ * See LICENSE for the license information
+
+ * -------------------------------------------------------------------------- */
+
+/*
+ * testDiscreteLookupDAG.cpp
```
Comment (varunagrawal) [inline gtsam/discrete/tests/testDiscreteLookupDAG.cpp:13]: Proper doxygen tags?

File: gtsam_unstable/discrete/CSP.h [hunk 11]
```diff
@@ -43,12 +43,6 @@ class GTSAM_UNSTABLE_EXPORT CSP : public DiscreteFactorGraph {
   //      return result;
   //    }
 
-  /// Find the best total assignment - can be expensive.
```
Comment (varunagrawal) [inline gtsam_unstable/discrete/CSP.h:46]: Should this be deprecated rather than removed?

File: gtsam_unstable/discrete/Scheduler.h [hunk 12]
```diff
@@ -147,9 +147,6 @@ class GTSAM_UNSTABLE_EXPORT Scheduler : public CSP {
   /** Eliminate, return a Bayes net */
   DiscreteBayesNet::shared_ptr eliminate() const;
 
-  /** Find the best total assignment - can be expensive */
-  DiscreteValues optimalAssignment() const;
```
Comment (varunagrawal) [inline gtsam_unstable/discrete/Scheduler.h:151]: Should this be deprecated instead of removed?

**Review by varunagrawal — COMMENTED** [review]
A few comments. I can address them if need be to save time.

**Comment by dellaert** [comment]
@varunagrawal sign off? Do you know what could be up with failing CI?

**Comment by varunagrawal** [comment]
This is the same issue that Fan had, it's a deprecation of the setup.py install command for python.
You can cherry pick this commit to fix it 2a023f7df

**Comment by dellaert** [comment]
Thanks. Done. Note it's not approved yet so I can't merge... Is there anything else?

**Review by varunagrawal — APPROVED** [review]
LGTM

**Comment by varunagrawal** [comment]
I was literally in the middle of the review :)

**Comment by dellaert** [comment]
Hmmm, after the cherry-pick, *10* runs are failing rather than 2… any ideas?

**Comment by dellaert** [comment]
@varunagrawal could you please explain more about your fix? Linux complains:
```
Disabling all use of wheels due to the use of --build-options / --global-options / --install-options
```
What was the issue you tried to solve? What article described the fix? You said “ deprecation of the setup.py install command for python”. Where was it=t deprecated?

finally, use of pip directly seems wrong. If anything, `$PYTHON$ -m pip` ?

**Comment by dellaert** [comment]
Also, why does not affect the entirety of GTSAM? If so, seems worthy of its own PR in GTSAM, with explanation of issue in PR description.

**Comment by varunagrawal** [comment]
The Python Packaging Authority has decided to deprecate direct builds via calling setup.py (aka `python setup.py install`) due to the need to enforce a standard for wheel builds. [More details here.](https://github.com/pypa/pip/issues/8368)

This effect first showed up on MacOS with the following error:
```
python3 setup.py install --user --prefix=
PACKAGES:  ['gtsam', 'gtsam_unstable', 'gtsam.utils', 'gtsam.examples', 'gtsam_unstable.tests', 'gtsam_unstable.examples']
running install
/usr/local/lib/python3.9/site-packages/setuptools/command/install.py:34: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
  warnings.warn(
/usr/local/lib/python3.9/site-packages/setuptools/command/easy_install.py:156: EasyInstallDeprecationWarning: easy_install command is deprecated. Use build and pip and other standards-based tools.
```

I myself use a `pip install -e .` for all my projects which basically asks pip to install the local directory in editable mode. Editable mode means the install reflects any local changes we make, making iterative development very easy.

We are now seeing the deprecation effects in Ubuntu as well, so yeah this is going to need a bit of refactoring of the CI. This is something that really annoys me about Github Actions is that they run environment upgrades pretty arbitrarily. I imagine this will effect the entirety of GTSAM going forward.

About using `pip` directly compared to `${PYTHON} -m pip`, they are equivalent. The `pip` binary is a tiny python script that imports the `pip` module and executes the command, making it easier. I can do the latter if you'd like.

**Comment by dellaert** [comment]
Awesome description above. Perhaps you can copy/paste some of the comment above in the (rather terse) comment of #1054 ? 

About GA: can't we specify whatever *exact* version we want ?

About pip: just heard that sometimes pip on path is a different version so #PYTHON -m pip ensures you have the pip version paired with the python version.

**Comment by varunagrawal** [comment]
We can always do `pip$PYTHON_VERSION` aka the exact pip version to match the python version as well, but I agree with using `python -m pip`.

**Pull request merged**